pkg/controller/seed-controller-manager/encryption/controller.go

```diff
@@ -0,0 +1,290 @@
+	versions          kubermatic.Versions
+}
+
+func Add(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+
+	numWorkers int,
+	workerName string,
+
+	userClusterConnProvider userClusterConnectionProvider,
+	seedGetter provider.SeedGetter,
+	configGetter provider.KubermaticConfigurationGetter,
+
+	versions kubermatic.Versions,
+	overwriteRegistry string,
+) error {
+	reconciler := &Reconciler{
+		log:                     log.Named(ControllerName),
+		Client:                  mgr.GetClient(),
+		userClusterConnProvider: userClusterConnProvider,
+		seedGetter:              seedGetter,
+		configGetter:            configGetter,
+		workerName:              workerName,
+
+		recorder: mgr.GetEventRecorderFor(ControllerName),
+
+		versions:          versions,
+		overwriteRegistry: overwriteRegistry,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
+		controllerutil.EnqueueClusterForNamespacedObject(mgr.GetClient()),
+		predicateutil.ByName(resources.EncryptionConfigurationSecretName),
+	); err != nil {
+		return fmt.Errorf("failed to create watcher for corev1.Secret: %w", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &appsv1.Deployment{}},
+		controllerutil.EnqueueClusterForNamespacedObject(mgr.GetClient()),
+		predicateutil.ByName(resources.ApiserverDeploymentName),
+	); err != nil {
+		return fmt.Errorf("failed to create watcher for appsv1.Deployment: %w", err)
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}, predicateutil.Factory(func(o ctrlruntimeclient.Object) bool {
+		cluster := o.(*kubermaticv1.Cluster)
+		encryptionEnabled := cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled
+		encryptionActive := cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue)
+		return cluster.Spec.Features[kubermaticv1.ClusterFeatureEncryptionAtRest] && (encryptionEnabled || encryptionActive)
+	}))
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("Could not find cluster")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	log = log.With("cluster", cluster.Name)
+
+	// replicate the predicate from above to make sure that reconcile loops triggered by apiserver and secret
+	// do not run unexpected reconciles.
+	if !(cluster.Spec.EncryptionConfiguration != nil && cluster.Spec.EncryptionConfiguration.Enabled) && !(cluster.Status.HasConditionValue(kubermaticv1.ClusterConditionEncryptionInitialized, corev1.ConditionTrue)) {
```

Comment:
You wanted nitpicks, you get them: `!(cluster.Status.Has(...))` is one pair of parenthesis too many.